### PR TITLE
Support for PHP 7.1

### PR DESCRIPTION
**Pull request by prolic**
- Support for PHP 7.1
- Remove deprecated AbstractAggregateRepositoryFactory (https://github.com/prooph/event-store/issues/171)
- uses custom prophecy branch (see https://github.com/phpspec/prophecy/pull/287)
- uses custom prooph/common branch ("prooph/common" : "dev-php_7_1 as 4.0")
- allow multiple event stores to be configured

### CODE REVIEW

**Comment by prolic** [comment]
```
2.97s$ ./vendor/bin/php-cs-fixer fix -v --diff --dry-run

Loaded config from "/home/travis/build/prooph/event-store/.php_cs".

....E..........................E............................................

Legend: ?-unknown, I-invalid file syntax, file ignored, .-no changes, F-fixed, E-error

Checked all files in 2.852 seconds, 8.000 MB memory used

Files that were not fixed due to internal error:

   1) Adapter/InMemoryAdapter.php

   2) Snapshot/Adapter/InMemoryAdapter.php
```

This is due a bug in php-cs-fixer with PHP 7.1

File: examples/quickstart.php [hunk 1]
```diff
@@ -98,6 +100,8 @@ function (\Prooph\Common\Event\ActionEvent $actionEvent) {
                 $recordedEvent->createdAt()->format('Y-m-d H:i:s')
             );
         }
+
+        return new \Prooph\Common\Event\DefaultListenerHandler($function);
```
Comment (Ocramius) [inline examples/quickstart.php]: Isn't this creating a new listener at each execution? Can lead to stack overflows?

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 2]
```diff
@@ -59,23 +61,14 @@ class ConfigurableAggregateTranslator implements AggregateTranslator
...
-     */
     public function __construct(
-        $identifierMethodName = null,
-        $versionMethodName = null,
-        $popRecordedEventsMethodName = null,
-        $replayEventsMethodsName = null,
-        $staticReconstituteFromHistoryMethodName = null,
-        $eventToMessageCallback = null,
-        $messageToEventCallback = null)
+        ?string $identifierMethodName = null,
```
Comment (Ocramius) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: When `= null` is provided, `?` is optional. Your decision whether to keep it though

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 3]
```diff
@@ -59,23 +61,14 @@ class ConfigurableAggregateTranslator implements AggregateTranslator
...
+        ?string $versionMethodName = null,
+        ?string $popRecordedEventsMethodName = null,
+        ?string $replayEventsMethodsName = null,
+        ?string $staticReconstituteFromHistoryMethodName = null,
+        ?callable $eventToMessageCallback = null,
+        ?callable $messageToEventCallback = null)
```
Comment (Ocramius) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: `)` is misaligned (PSR-2)

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 4]
```diff
@@ -103,12 +96,10 @@ public function __construct(
         }
 
         if (null !== $eventToMessageCallback) {
-            Assertion::true(is_callable($eventToMessageCallback), 'EventToMessage callback needs to be a callable');
```
Comment (Ocramius) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: BANG! \o/

File: src/Container/EventStoreFactory.php [hunk 5]
```diff
@@ -114,7 +112,7 @@ public function mandatoryOptions()
     /**
      * @inheritdoc
      */
-    public function defaultOptions()
+    public function defaultOptions(): array
```
Comment (Ocramius) [inline src/Container/EventStoreFactory.php]: These could be moved to VOs in the future

File: src/EventStore.php [hunk 6]
```diff
@@ -264,7 +238,7 @@ public function loadEventsByMetadataFrom(StreamName $streamName, array $metadata
      * @return CompositeIterator
      * @throws Exception\InvalidArgumentException
      */
-    public function replay(array $streamNames, DateTimeInterface $since = null, array $metadatas = null)
+    public function replay(array $streamNames, DateTimeInterface $since = null, array $metadatas = null): CompositeIterator
```
Comment (Ocramius) [inline src/EventStore.php]: Why `CompositeIterator`? Does it make sense from a consumer perspective? Also, the docblock is not needed anymore.

File: src/EventStore.php [hunk 7]
```diff
@@ -297,9 +271,7 @@ public function replay(array $streamNames, DateTimeInterface $since = null, arra
 
     /**
      * @param callable $callable
-     *
```
Comment (Ocramius) [inline src/EventStore.php]: Note: spacing between dockblock tag types should be kept overall

File: examples/quickstart.php (hunk 1 above)
Comment (Ocramius) [inline examples/quickstart.php]: Good point. Still very confusing though.

File: src/Aggregate/ConfigurableAggregateTranslator.php (hunk 4 above)
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: What? There is a callable typehint.

File: src/Aggregate/ConfigurableAggregateTranslator.php (hunk 2 above)
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: done

File: src/Container/EventStoreFactory.php (hunk 5 above)
Comment (prolic) [inline src/Container/EventStoreFactory.php]: Even more classes?

File: src/EventStore.php (hunk 6 above)
Comment (prolic) [inline src/EventStore.php]: fixed

File: src/EventStore.php (hunk 7 above)
Comment (prolic) [inline src/EventStore.php]: ok

File: src/Container/EventStoreFactory.php (hunk 5 above)
Comment (Ocramius) [inline src/Container/EventStoreFactory.php]: Eventually, doesn't need to happen now, but it's indeed to be considered

File: examples/quickstart.php [hunk 8]
```diff
@@ -82,7 +84,7 @@
  */
 $eventStore->getActionEventEmitter()->attachListener(
     'commit.post', //Most of the event store methods provide pre and post hooks
-    function (\Prooph\Common\Event\ActionEvent $actionEvent) {
+    $function = function (\Prooph\Common\Event\ActionEvent $actionEvent) use (&$function) {
```
Comment (oqq) [inline examples/quickstart.php]: Missing return type hint. Should be :\Prooph\Common\Event\DefaultListenerHandler

Comment (oqq) [inline examples/quickstart.php]: Is that var "$function" used?

File: src/Aggregate/AggregateRepository.php [hunk 9]
```diff
@@ -147,9 +142,10 @@ public function addAggregateRoot($eventSourcedAggregateRoot)
      * Returns null if no stream events can be found for aggregate root otherwise the reconstituted aggregate root
      *
      * @param string $aggregateId
+     *
      * @return null|object
      */
-    public function getAggregateRoot($aggregateId)
+    public function getAggregateRoot(string $aggregateId)
     {
         Assertion::string($aggregateId, 'AggregateId needs to be string');
```
Comment (oqq) [inline src/Aggregate/AggregateRepository.php]: unnecessary assert call

File: src/Aggregate/AggregateTranslator.php [hunk 10]
```diff
@@ -23,32 +25,40 @@
...
      * @param object $eventSourcedAggregateRoot
+     *
      * @return string
      */
-    public function extractAggregateId($eventSourcedAggregateRoot);
+    public function extractAggregateId($eventSourcedAggregateRoot): string;
 
     /**
      * @param AggregateType $aggregateType
+     *
```
Comment (oqq) [inline src/Aggregate/AggregateTranslator.php]: empty line between equal annotation tag types

File: src/Aggregate/AggregateTranslator.php [hunk 11]
```diff
@@ -23,32 +25,40 @@
...
      * @return Message[]
      */
-    public function extractPendingStreamEvents($eventSourcedAggregateRoot);
+    public function extractPendingStreamEvents($eventSourcedAggregateRoot): array;
 
     /**
-     * @param $anEventSourcedAggregateRoot
+     * @param object $eventSourcedAggregateRoot
+     *
```
Comment (oqq) [inline src/Aggregate/AggregateTranslator.php]: empty line between equal annotation tag types

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 12]
```diff
@@ -154,8 +148,11 @@ public function extractAggregateVersion($eventSourcedAggregateRoot)
 
     /**
      * @param AggregateType $aggregateType
+     *
```
Comment (oqq) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: empty line between equal annotation tag types

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 13]
```diff
@@ -257,10 +256,14 @@ public function extractPendingStreamEvents($eventSourcedAggregateRoot)
 
     /**
      * @param object $eventSourcedAggregateRoot
+     *
```
Comment (oqq) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: empty line between equal annotation tag types

File: src/Container/Aggregate/AggregateRepositoryFactory.php [hunk 14]
```diff
@@ -94,27 +86,27 @@ public function __invoke(ContainerInterface $container)
...
+        $oneStreamPerAggregate = (bool) ($config['one_stream_per_aggregate'] ?? false);
 
         return new $repositoryClass($eventStore, $aggregateType, $aggregateTranslator, $snapshotStore, $streamName, $oneStreamPerAggregate);
     }
 
     /**
      * @inheritdoc
      */
-    public function dimensions()
+    public function dimensions(): array
```
Comment (oqq) [inline src/Container/Aggregate/AggregateRepositoryFactory.php]: interop-config interfaces are defined without return type hints. Maybe we should omit they here.

File: src/Container/Aggregate/AggregateRepositoryFactory.php [hunk 15]
```diff
@@ -94,27 +86,27 @@ public function __invoke(ContainerInterface $container)
...
     {
         return ['prooph', 'event_store'];
     }
 
     /**
      * @inheritdoc
      */
-    public function mandatoryOptions()
+    public function mandatoryOptions(): array
```
Comment (oqq) [inline src/Container/Aggregate/AggregateRepositoryFactory.php]: ...and here

File: src/Snapshot/Snapshot.php [hunk 16]
```diff
@@ -47,16 +49,20 @@
 
     /**
      * @param AggregateType $aggregateType
+     *
```
Comment (oqq) [inline src/Snapshot/Snapshot.php]: empty line between equal annotation tag types

File: src/Util/CompositeIterator.php [hunk 17]
```diff
@@ -32,7 +34,9 @@
 
     /**
      * @param Iterator[]|IteratorAggregate[] $iterators
+     *
```
Comment (oqq) [inline src/Util/CompositeIterator.php]: empty line between equal annotation tag types

File: tests/Metadata/MetadataEnricherPluginTest.php [hunk 18]
```diff
@@ -37,13 +40,15 @@ public function it_attaches_itself_to_event_store_events()
 
         $eventStore->getActionEventEmitter()->willReturn($eventEmitter);
         $eventEmitter->attachListener('create.pre', Argument::any(), -1000)->will(
-            function ($args) use (&$createStreamListener) {
+            $function = function ($args) use (&$createStreamListener, &$function) {
```
Comment (oqq) [inline tests/Metadata/MetadataEnricherPluginTest.php]: No return type hint provided. Should be : DefaultListenerHandler

File: tests/Metadata/MetadataEnricherPluginTest.php [hunk 19]
```diff
@@ -37,13 +40,15 @@ public function it_attaches_itself_to_event_store_events()
...
                 $createStreamListener = $args[1];
+                return new DefaultListenerHandler($function);
             }
         );
         $eventEmitter->attachListener('appendTo.pre', Argument::any(), -1000)->will(
-            function ($args) use (&$appendToStreamListener) {
+            $function = function ($args) use (&$appendToStreamListener, &$function) {
```
Comment (oqq) [inline tests/Metadata/MetadataEnricherPluginTest.php]: No return type hint provided. Should be : DefaultListenerHandler

File: examples/quickstart.php (hunk 8 above)
Comment (prolic) [inline examples/quickstart.php]: yes in the return statement

File: src/Aggregate/AggregateRepository.php (hunk 9 above)
Comment (prolic) [inline src/Aggregate/AggregateRepository.php]: fixed

File: src/Aggregate/ConfigurableAggregateTranslator.php (hunk 13 above)
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: fixed

File: src/Aggregate/AggregateTranslator.php (hunk 11 above)
Comment (prolic) [inline src/Aggregate/AggregateTranslator.php]: fixed

File: src/Aggregate/ConfigurableAggregateTranslator.php (hunk 12 above)
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: fixed

File: tests/Metadata/MetadataEnricherPluginTest.php (hunk 19 above)
Comment (prolic) [inline tests/Metadata/MetadataEnricherPluginTest.php]: fixed

File: tests/Metadata/MetadataEnricherPluginTest.php (hunk 18 above)
Comment (prolic) [inline tests/Metadata/MetadataEnricherPluginTest.php]: fixed

File: src/Snapshot/Snapshot.php (hunk 16 above)
Comment (prolic) [inline src/Snapshot/Snapshot.php]: fixed

File: src/Aggregate/ConfigurableAggregateTranslator.php (hunk 4 above)
Comment (codeliner) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: unbelievable! @Ocramius is everywhere.

**Comment by Ocramius** [comment]
💃

**Pull request merged**